### PR TITLE
Use game state structure for scenario construction globals

### PR DESCRIPTION
**Pull request by memellis**
My first attempt with moving the following related globals:

static loco_global<uint8_t[8], 0x0525F7A> _scenarioBridges;
static loco_global<uint8_t[8], 0x0525F82> _scenarioTrainStations;
static loco_global<uint8_t[8], 0x0525F8A> _scenarioTrackMods;
static loco_global<uint8_t[8], 0x0525F9A> _scenarioRoadStations;
static loco_global<uint8_t[8], 0x0525FA2> _scenarioRoadMods;

I moved these into a new ScenarioConstruction.h, with a corresponding ScenarioConstruction.cpp.

I don't think I've addressed the "With only one route to access particular members of the game state" yet.
Very happy to change following review comments.

### CODE REVIEW

File: src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp [hunk 1]
```diff
@@ -136,10 +136,10 @@ namespace OpenLoco::Ui::Windows::Construction::Construction
...
     void reset()
     {
-        std::fill(std::begin(_scenarioSignals), std::end(_scenarioSignals), 0xFF);
-        std::fill(std::begin(_scenarioBridges), std::end(_scenarioBridges), 0xFF);
-        std::fill(std::begin(_scenarioTrainStations), std::end(_scenarioTrainStations), 0xFF);
-        std::fill(std::begin(_scenarioTrackMods), std::end(_scenarioTrackMods), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().signals), std::end(Scenario::getConstruction().signals), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().bridges), std::end(Scenario::getConstruction().bridges), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().trainStations), std::end(Scenario::getConstruction().trainStations), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().trackMods), std::end(Scenario::getConstruction().trackMods), 0xFF);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp:142]: Maybe we should introduce a reset function for these in the `Scenario::Construction` struct you've introduced, resetting the entire struct to 0xFF.

File: src/OpenLoco/src/ScenarioConstruction.cpp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#include "./GameState.h"
+#include "./ScenarioConstruction.h"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/ScenarioConstruction.cpp:2]:
```suggestion
#include "ScenarioConstruction.h"
#include "GameState.h"
```

**Comment by AaronVanGeffen** [comment]
This includes ImageId-related commits from #1903, which have since been merged into master. Could you rebase on top of master, dropping those commits?

File: src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp (hunk 1 above)
Comment (memellis) [reply src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp:142]: Agreed. Shall I add reset lines for roadStations and roadMods and are which is missing in ConstructionTab reset function?

**Comment by memellis** [comment]
I've hopefully rebased dropping the ImageId-related commits from https://github.com/OpenLoco/OpenLoco/pull/1903.

**Comment by memellis** [comment]
Not sorted the rebasing to drop the ImageId-related commits from #1903, yet. I'll try again tomorrow.

File: src/OpenLoco/src/ScenarioConstruction.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        return getGameState().scenarioConstruction;
+    }
+
+    void resetConstruction()
+    {
+        std::fill(std::begin(Scenario::getConstruction().signals), std::end(Scenario::getConstruction().signals), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().bridges), std::end(Scenario::getConstruction().bridges), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().trainStations), std::end(Scenario::getConstruction().trainStations), 0xFF);
+        std::fill(std::begin(Scenario::getConstruction().trackMods), std::end(Scenario::getConstruction().trackMods), 0xFF);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/ScenarioConstruction.cpp:16]:
```suggestion
        auto& construction = getConstruction();
        std::fill(std::begin(construction.signals), std::end(construction.signals), 0xFF);
        std::fill(std::begin(construction.bridges), std::end(construction.bridges), 0xFF);
        std::fill(std::begin(construction.trainStations), std::end(construction.trainStations), 0xFF);
        std::fill(std::begin(construction.trackMods), std::end(construction.trackMods), 0xFF);
```

**Review by AaronVanGeffen — APPROVED** [review]
Looks good to me, though I'm not quite sure `Scenario::Construction` is the right name for this. Can't think of anything better either, though.

@duncanspumpkin Any opinions?

**Comment by duncanspumpkin** [comment]
> Looks good to me, though I'm not quite sure `Scenario::Construction` is the right name for this. Can't think of anything better either, though.
> 
> @duncanspumpkin Any opinions?

Yeah can't think of a better name atm. Would be good to investigate that unnamed field as well

**Comment by memellis** [comment]
A quick look (with the help of IDA) for the unamed variable:

- uint8_t var_17A[8];       // 0x00017A (0x00525F92)

Location is 0x00525F92 8 bytes are initialised to 0xFF in:
'''
.text:0047598A loc_47598A:                             ; CODE XREF: OpenLoco::Scenario::sub_475988(void)+1B↓j
.text:0047598A                 mov     byte_525F92[edi], 0FFh
.text:00475991                 mov     _scenarioRoadStations[edi], 0FFh
.text:00475998                 mov     _scenarioRoadMods[edi], 0FFh
.text:0047599F                 inc     edi
.text:004759A0                 cmp     edi, 8
.text:004759A3                 jb      short loc_47598A
.text:004759A5                 retn
.text:004759A5 ?sub_475988@Scenario@OpenLoco@@YAXXZ endp
'''

0x00525F92 is manipulated around this bit of assembly:
'''
.text:004A1822 loc_4A1822:                             ; CODE XREF: cursor_4C74BB-25CA0↑j
.text:004A1822                 mov     _lastSelectedStationType, al
.text:004A1827                 mov     al, _trackType
.text:004A182C                 and     al, 7Fh
.text:004A182E                 mov     edi, offset _bridgeList
.text:004A1833                 call    ?refreshBridgeList@Common@Construction@Windows@Ui@OpenLoco@@YAXPEAEEW4TransportMode@5@@Z ; OpenLoco::Ui::Windows::Construction::Common::refreshBridgeList(uchar *,uchar,OpenLoco::TransportMode)
.text:004A1838                 movzx   edi, _trackType
.text:004A183F                 and     di, 0FF7Fh
.text:004A1844                 mov     al, byte_525F92[edi]
.text:004A184A                 cmp     al, 0FFh
.text:004A184C                 jnz     short loc_4A1853
.text:004A184E                 mov     al, _bridgeList
.text:004A1853
.text:004A1853 loc_4A1853:                             ; CODE XREF: cursor_4C74BB-25C6F↑j
.text:004A1853                 mov     _lastSelectedBridge, al
.text:004A1858                 mov     edx, dword_1135F4A
.text:004A185E                 test    byte ptr [edx+4], 80h
.text:004A1862                 jz      short loc_4A186F
.text:004A1864                 mov     al, [edx+6]
.text:004A1867                 shr     al, 5
.text:004A186A                 mov     _lastSelectedBridge, al
'''

So look like register al gets loaded with a value of index 0x00525F92 +_tracktype and compared to 0FFh.
- If register al=0FFh then set al with _bridgeList 
- Otherwise set _lastSelectedBridge with regsiter al (which I assumes contains the last selected bridge value)

 In conclusion I think:
- 0x00525F92[8] is only ever set to 0FFH
- 0x00525F92[8] looks like its was associated with road bridges based on track type
- But the bridges[8] array can deal with both road and bridges leaving 0x00525F92[8] redundant?

Here is the relevant code in OpenLoco/src/Windows/Common.cpp from line 361: 

Common::refreshBridgeList(_bridgeList, _trackType, TransportMode::road);

auto lastBridge = Scenario::getConstruction().bridges[(_trackType & ~(1ULL << 7))];

if (lastBridge == 0xFF)
    lastBridge = _bridgeList[0];

_lastSelectedBridge = lastBridge;
if (copyElement->hasBridge())
{
       _lastSelectedBridge = copyElement->bridge();
}

Note: this is very much guess work!

**Comment by duncanspumpkin** [comment]
Yeah doesn't look like it has any value other than 0xFF and it is unused. Thats good. Could you add a note on the field that it is "not used always filled with 0xFF"

**Review by duncanspumpkin — APPROVED** [review]
